library/pkcs7.c

```diff
@@ -0,0 +1,525 @@
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_FORMAT + ret );
+
+    /* Get version of signed data */
+    ret = pkcs7_get_version( &p, end, &signed_data->version );
+    if( ret != 0 )
+        return( ret );
+
+    /* If version != 1, return invalid version */
+    if( signed_data->version != MBEDTLS_PKCS7_SUPPORTED_VERSION ) {
+        return( MBEDTLS_ERR_PKCS7_INVALID_VERSION );
+    }
+
+    /* Get digest algorithm */
+    ret = pkcs7_get_digest_algorithm_set( &p, end,
+            &signed_data->digest_alg_identifiers );
+    if( ret != 0 ) {
+        return( ret );
+    }
+
+    ret = mbedtls_oid_get_md_alg( &signed_data->digest_alg_identifiers, &md_alg );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_ALG + ret );
+
+    /* Do not expect any content */
+    ret = pkcs7_get_content_info_type( &p, end, &signed_data->content.oid );
+    if( ret != 0 )
+        return( ret );
+
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DATA, &signed_data->content.oid ) ) {
+        return( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO ) ;
+    }
+
+    p = p + signed_data->content.oid.len;
+
+    /* Look for certificates, there may or may not be any */
+    ret = pkcs7_get_next_content_len( &p, end, &len );
+    if( ret == 0 ) {
+        mbedtls_x509_crt_init( &signed_data->certs );
+        ret = pkcs7_get_certificates( &p, len, &signed_data->certs );
+        if( ret != 0 )
+            return( ret ) ;
+
+      p = p + len;
+    }
+
+    /* TODO: optional CRLs go here */
+
+    /* Get signers info */
+    ret = pkcs7_get_signers_info_set( &p, end, &signed_data->signers );
+    if( ret != 0 )
+        return( ret );
+
+    return( ret );
+}
+
+int mbedtls_pkcs7_parse_der( const unsigned char *buf, const int buflen,
+        mbedtls_pkcs7 *pkcs7 )
+{
+    unsigned char *start;
+    unsigned char *end;
+    size_t len = 0;
+    int ret;
+
+    /* use internal buffer for parsing */
+    start = ( unsigned char * )buf;
+    end = start + buflen;
+
+    if(!pkcs7)
+        return( MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA );
+
+    ret = pkcs7_get_content_info_type( &start, end, &pkcs7->content_type_oid );
+    if( ret != 0 )
+        goto out;
+
+    if( ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENVELOPED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_SIGNED_AND_ENVELOPED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DIGESTED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid ) )
+    {
+        ret =  MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE;
+        goto out;
+    }
+
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_SIGNED_DATA, &pkcs7->content_type_oid ) )
+    {
+        ret = MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA;
```

Comment:
Shouldn't the usage of errors be the other way around? Returning `MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA` for the previous check, and `FEATURE_UNAVAILABLE` in the present one?